File: ai-rag-android/app/src/main/java/net/apperdashery/demo/ai_rag_android/store/AppReducer.kt
package net.apperdashery.demo.ai_rag_android.store

fun appReducer(state: AppState, action: AppAction): AppState =
    when (action) {
        is AppAction.Osha -> state.copy(osha = oshaReducer(state.osha, action.action))
    }

fun oshaReducer(state: OshaState, action: OshaAction): OshaState =
    when (action) {
        is OshaAction.AskExpertTapped -> state.copy(
            isLoading = true,
            response = null,
            errorMessage = null
        )

        is OshaAction.ResponseReceived -> state.copy(
            isLoading = false,
            response = action.response,
            errorMessage = null,
            route = OshaRoute.Response
        )

        is OshaAction.RequestFailed -> state.copy(
            isLoading = false,
            response = null,
            errorMessage = action.message,
            route = OshaRoute.Response
        )

        is OshaAction.DismissedResponse -> state.copy(
            response = null,
            errorMessage = null,
            route = null
        )

        is OshaAction.NavigationHandled -> state.copy(route = null)
    }

File: ai-rag-app/ai-rag-app/Store/AppReducer.swift
//
//  AppReducer.swift
//  ai-rag-app
//
//  Created by Graham Diehl on 8/26/26.
//

import Foundation

func appReducer(_ state: AppState, _ action: AppAction) -> AppState {
    var newState = state
    switch action {
    case .osha(let oshaAction):
        newState.osha = oshaReducer(state.osha, oshaAction)
    }
    return newState
}

func oshaReducer(_ state: OshaState, _ action: OshaAction) -> OshaState {
    var newState = state
    switch action {
    case .askExpertTapped:
        newState.isLoading = true
        newState.response = nil
        newState.errorMessage = nil

    case .responseReceived(let response):
        newState.isLoading = false
        newState.response = response
        newState.errorMessage = nil
        newState.route = [.response]

    case .requestFailed(let message):
        newState.isLoading = false
        newState.response = nil
        newState.errorMessage = message
        newState.route = [.response]

    case .dismissedResponse:
        newState.route = []
    }
    return newState
}
